From 076d48c07fe529cbd4c10e599174ce2ea4a7c967 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Stephan=20D=C3=BCnner?= <stephan.duenner@bytesatwork.ch>
Date: Tue, 15 Oct 2019 13:30:19 +0200
Subject: [PATCH] byteengine: Add qSPI interface support for booting from
 SPI-NOR flash

The "Boot select" switch on byteDEVKIT needs to be set to "NOR" for this to
work.
---
 .../dts/stm32mp157c-bytedevkit-u-boot.dtsi    | 28 +++++++++++
 arch/arm/dts/stm32mp157c-bytedevkit.dts       |  1 +
 arch/arm/dts/stm32mp157c-byteengine-m5.dtsi   | 50 +++----------------
 configs/stm32mp157_bytedevkit_defconfig       | 14 ++++--
 include/configs/bytedevkit.h                  | 15 ++++++
 5 files changed, 61 insertions(+), 47 deletions(-)

diff --git a/arch/arm/dts/stm32mp157c-bytedevkit-u-boot.dtsi b/arch/arm/dts/stm32mp157c-bytedevkit-u-boot.dtsi
index b953975dc9..c59b589cd7 100644
--- a/arch/arm/dts/stm32mp157c-bytedevkit-u-boot.dtsi
+++ b/arch/arm/dts/stm32mp157c-bytedevkit-u-boot.dtsi
@@ -11,6 +11,7 @@
 	aliases {
 		i2c3 = &i2c4;
 		mmc0 = &sdmmc1;
+		spi0 = &qspi;
 	};
 };
 
@@ -144,6 +145,33 @@
 	u-boot,dm-pre-reloc;
 };
 
+
+&flash0 {
+	compatible = "spi-flash";
+	u-boot,dm-spl;
+};
+
+&qspi {
+	u-boot,dm-spl;
+};
+
+&qspi_clk_pins_a {
+	u-boot,dm-spl;
+	pins {
+		u-boot,dm-spl;
+	};
+};
+
+&qspi_bk1_pins_a {
+	u-boot,dm-spl;
+	pins1 {
+		u-boot,dm-spl;
+	};
+	pins2 {
+		u-boot,dm-spl;
+	};
+};
+
 &uart4_pins_a {
 	u-boot,dm-pre-reloc;
 	pins1 {
diff --git a/arch/arm/dts/stm32mp157c-bytedevkit.dts b/arch/arm/dts/stm32mp157c-bytedevkit.dts
index dd2697be8a..4deb8667bb 100644
--- a/arch/arm/dts/stm32mp157c-bytedevkit.dts
+++ b/arch/arm/dts/stm32mp157c-bytedevkit.dts
@@ -13,6 +13,7 @@
 
 	aliases {
 		serial0 = &uart4;
+		spi0 = &qspi;
 	};
 
 	chosen {
diff --git a/arch/arm/dts/stm32mp157c-byteengine-m5.dtsi b/arch/arm/dts/stm32mp157c-byteengine-m5.dtsi
index 6ce9c3b1ce..688db8d75a 100644
--- a/arch/arm/dts/stm32mp157c-byteengine-m5.dtsi
+++ b/arch/arm/dts/stm32mp157c-byteengine-m5.dtsi
@@ -131,43 +131,6 @@
 				 <STM32_PINMUX('G', 6, ANALOG)>; /* SDMMC2_CMD */
 		};
 	};
-
-	qspi_pins: qspi-bk1-0 {
-		pins1 {
-			pinmux = <STM32_PINMUX('F', 8, AF10)>, /* QSPI_BK1_IO0 */
-				 <STM32_PINMUX('F', 9, AF10)>, /* QSPI_BK1_IO1 */
-				 <STM32_PINMUX('F', 7, AF9)>, /* QSPI_BK1_IO2 */
-				 <STM32_PINMUX('F', 6, AF9)>; /* QSPI_BK1_IO3 */
-			bias-disable;
-			drive-push-pull;
-			slew-rate = <1>;
-		};
-
-		pins2 {
-			pinmux = <STM32_PINMUX('B', 6, AF10)>; /* QSPI_BK1_NCS */
-			bias-pull-up;
-			drive-push-pull;
-			slew-rate = <1>;
-		};
-
-		pins3 {
-			pinmux = <STM32_PINMUX('F', 10, AF9)>; /* QSPI_CLK */
-			bias-disable;
-			drive-push-pull;
-			slew-rate = <3>;
-		};
-	};
-
-	qspi_sleep_pins: qspi-bk1-sleep-0 {
-		pins {
-			pinmux = <STM32_PINMUX('F', 8, ANALOG)>, /* QSPI_BK1_IO0 */
-				 <STM32_PINMUX('F', 9, ANALOG)>, /* QSPI_BK1_IO1 */
-				 <STM32_PINMUX('F', 7, ANALOG)>, /* QSPI_BK1_IO2 */
-				 <STM32_PINMUX('F', 6, ANALOG)>, /* QSPI_BK1_IO3 */
-				 <STM32_PINMUX('B', 6, ANALOG)>, /* QSPI_BK1_NCS */
-				 <STM32_PINMUX('F', 10, ANALOG)>; /* QSPI_CLK */
-		};
-	};
 };
 
 &pinctrl_z {
@@ -366,17 +329,18 @@
 /* NOR Flash */
 &qspi {
 	pinctrl-names = "default", "sleep";
-	pinctrl-0 = <&qspi_pins>;
-	pinctrl-1 = <&qspi_sleep_pins>;
+	pinctrl-0 = <&qspi_clk_pins_a &qspi_bk1_pins_a &qspi_bk2_pins_a>;
+	pinctrl-1 = <&qspi_clk_sleep_pins_a &qspi_bk1_sleep_pins_a &qspi_bk2_sleep_pins_a>;
+	reg = <0x58003000 0x1000>, <0x70000000 0x4000000>;
 	#address-cells = <1>;
 	#size-cells = <0>;
 	status = "okay";
 
-	flash0: w25n01g@0 {
-		compatible = "st,m25pe40", "jedec,spi-nor";
+	flash0: w25q128jv@0 {
+		compatible = "jedec,spi-nor";
 		reg = <0>;
-		spi-rx-bus-width = <2>;
-		spi-max-frequency = <10000000>;
+		spi-rx-bus-width = <4>;
+		spi-max-frequency = <108000000>;
 		#address-cells = <1>;
 		#size-cells = <1>;
 	};
diff --git a/configs/stm32mp157_bytedevkit_defconfig b/configs/stm32mp157_bytedevkit_defconfig
index 0a039a191a..f329d01533 100644
--- a/configs/stm32mp157_bytedevkit_defconfig
+++ b/configs/stm32mp157_bytedevkit_defconfig
@@ -13,6 +13,7 @@ CONFIG_SUPPORT_RAW_INITRD=y
 CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_USE_PARTITION=y
 CONFIG_SYS_MMCSD_RAW_MODE_U_BOOT_PARTITION=3
 CONFIG_SPL_I2C_SUPPORT=y
+CONFIG_SPL_MTD_SUPPORT=y
 CONFIG_SPL_POWER_SUPPORT=y
 CONFIG_HUSH_PARSER=y
 CONFIG_SYS_PROMPT="byteDEVKIT> "
@@ -26,15 +27,18 @@ CONFIG_CMD_BOOTZ=y
 CONFIG_CMD_MEMINFO=y
 CONFIG_CMD_MEMTEST=y
 CONFIG_CMD_ADC=y
+CONFIG_CMD_BIND=y
 CONFIG_CMD_CLK=y
 CONFIG_CMD_FUSE=y
 CONFIG_CMD_GPIO=y
 CONFIG_CMD_GPT=y
 CONFIG_CMD_I2C=y
 CONFIG_CMD_MMC=y
+CONFIG_CMD_MTD=y
 CONFIG_CMD_PART=y
 CONFIG_CMD_REMOTEPROC=y
 CONFIG_CMD_SF=y
+CONFIG_CMD_SPI=y
 CONFIG_CMD_DHCP=y
 CONFIG_CMD_MII=y
 CONFIG_CMD_PING=y
@@ -48,6 +52,9 @@ CONFIG_CMD_EXT2=y
 CONFIG_CMD_EXT4=y
 CONFIG_CMD_FAT=y
 CONFIG_CMD_FS_GENERIC=y
+CONFIG_CMD_MTDPARTS=y
+CONFIG_MTDIDS_DEFAULT="nor0=w25q128-flash.0"
+CONFIG_MTDPARTS_DEFAULT="mtdparts_nor1=256k(fsbl1),256k(fsbl2),2m(ssbl),256k(u-boot-env),-(nor_user)"
 # CONFIG_SPL_DOS_PARTITION is not set
 CONFIG_ISO_PARTITION=y
 CONFIG_DEFAULT_DEVICE_TREE="stm32mp157c-bytedevkit"
@@ -59,14 +66,12 @@ CONFIG_DM_MAILBOX=y
 CONFIG_STM32_IPCC=y
 CONFIG_DM_MMC=y
 CONFIG_STM32_SDMMC2=y
+CONFIG_MTD=y
 CONFIG_DM_SPI_FLASH=y
 CONFIG_SPI_FLASH=y
-CONFIG_SPI_FLASH_BAR=y
-CONFIG_SPI_FLASH_MACRONIX=y
-CONFIG_SPI_FLASH_SPANSION=y
-CONFIG_SPI_FLASH_STMICRO=y
 CONFIG_SPI_FLASH_WINBOND=y
 # CONFIG_SPI_FLASH_USE_4K_SECTORS is not set
+CONFIG_SPI_FLASH_MTD=y
 CONFIG_PINCONF=y
 # CONFIG_SPL_PINCTRL_FULL is not set
 CONFIG_PINCTRL_STMFX=y
@@ -83,4 +88,5 @@ CONFIG_SERIAL_RX_BUFFER=y
 CONFIG_SPI=y
 CONFIG_DM_SPI=y
 CONFIG_STM32_QSPI=y
+# CONFIG_REGEX is not set
 # CONFIG_EFI_LOADER is not set
diff --git a/include/configs/bytedevkit.h b/include/configs/bytedevkit.h
index 3c5275b38c..7273fae1d0 100644
--- a/include/configs/bytedevkit.h
+++ b/include/configs/bytedevkit.h
@@ -90,6 +90,8 @@
 #define CONFIG_SYS_SPI_U_BOOT_OFFS	0x80000
 #endif
 
+#define CONFIG_SYS_MAX_FLASH_BANKS	1
+
 /* FILE SYSTEM */
 
 #if defined(CONFIG_STM32_QSPI) || defined(CONFIG_NAND_STM32_FMC2)
@@ -124,6 +126,11 @@
 	"stdout=serial\0" \
 	"stderr=serial\0" \
 	"bootdelay=1\0" \
+	"spl_file=u-boot-spl.stm32\0" \
+	"uboot_file=u-boot.img\0" \
+	"uboot_offset=0x80000\0" \
+	"loadaddr=0xc1000000\0" \
+	"spl_uboot_size=0x280000\0" \
 	"kernel_addr_r=0xc2000000\0" \
 	"fdt_addr_r=0xc4000000\0" \
 	"scriptaddr=0xc4100000\0" \
@@ -137,6 +144,14 @@
 	"default_args=rootwait rw vt.global_cursor_default=0 consoleblank=0\0" \
 	"mmc_args=setenv bootargs ${default_args} console=${console} " \
 		"root=${mmc_root} ${bootargs_append}; \0" \
+	"update_spiflash=echo Updating SPI Flash ...; " \
+		"sf probe 0; " \
+		"sf erase 0 +${spl_uboot_size}; " \
+		"mmc rescan; " \
+		"ext4load mmc 0:4 ${loadaddr} ${spl_file} || exit; "\
+		"sf write ${loadaddr} 0 ${filesize}; " \
+		"ext4load mmc 0:4 ${loadaddr} ${uboot_file} || exit; " \
+		"sf write ${loadaddr} ${uboot_offset} ${filesize};\0" \
 	MMC_BOOT_CMD
 
 #endif /* ifndef CONFIG_SPL_BUILD */
-- 
2.20.1

